### PR TITLE
Number of pages not allow negative values

### PR DESCRIPTION
**Pull request by amit1307**
## Summary of change

Add validation for number of pages.

## Additional context

image

## Related issue

https://github.com/knjk04/book-project/issues/184

### CODE REVIEW

**Comment by knjk04** [comment]
Thanks a lot for this. 

A couple of things before I review it in full:

1) Could you resolve the merge conflicts?

2) Could you change the binder error message to the more user-friendly "There must be at least one page in the book"

**Comment by amit1307** [comment]
> Thanks a lot for this.
> 
> A couple of things before I review it in full:
> 
> 1. Could you resolve the merge conflicts?
> 2. Could you change the binder error message to the more user-friendly "There must be at least one page in the book"

Thanks, both done.

File: src/main/java/com/karankumar/bookproject/ui/FormErrors.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package com.karankumar.bookproject.ui;
+
+public class FormErrors {
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/FormErrors.java:3]: Thanks for refactoring this out into a new class. It's quite nice this way.

Since these error messages will only be used in the `BookForm`, could you move it into the ui/book package?

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java [hunk 2]
```diff
@@ -47,6 +48,8 @@
 import java.util.Objects;
 import java.util.logging.Level;
 
+import static com.karankumar.bookproject.ui.FormErrors.*;
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/BookForm.java:51]: We're not using wildcard imports (see https://google.github.io/styleguide/javaguide.html#s3.3.1-wildcard-imports)

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java [hunk 3]
```diff
@@ -158,47 +161,43 @@ private void closeForm() {
      * Binds the form fields and defines custom validators where necessary
      */
     private void bindFormFields() {
-        final String AFTER_TODAY_PREFIX = "The date you";
-        final String AFTER_TODAY_SUFFIX = "reading the book cannot be after today's date.";
-
         binder.forField(bookTitle)
-              .asRequired("Please provide a book title")
-              .bind(Book::getTitle, Book::setTitle);
+                .asRequired(BOOK_TITLE_ERROR)
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/BookForm.java:165]: Could you ensure the period aligns with the previous method call (`forField()`)?

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java [hunk 4]
```diff
@@ -158,47 +161,43 @@ private void closeForm() {
...
         binder.forField(dateFinishedReading)
-              .withValidator(endDate -> !(endDate != null && dateStartedReading.getValue() != null
-                              && endDate.isBefore(dateStartedReading.getValue())),
-                      "The date you finished reading the book cannot be earlier than the date you"
-                              + " started reading the book")
-              .withValidator(endDate -> !(endDate != null && endDate.isAfter(LocalDate.now())),
-                      AFTER_TODAY_PREFIX + " finished " + AFTER_TODAY_SUFFIX)
-              .bind(Book::getDateFinishedReading, Book::setDateFinishedReading);
+                .withValidator(endDatePredicate(), FINISH_DATE_ERROR)
+                .withValidator(datePredicate(), String.format(AFTER_TODAY_ERROR, "finished"))
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/BookForm.java:184]: I like how you've used String.format() here!

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java [hunk 5]
```diff
@@ -158,47 +161,43 @@ private void closeForm() {
...
-              .bind(Book::getGenre, Book::setGenre);
+                .bind(Book::getGenre, Book::setGenre);
         binder.forField(rating)
-              .withConverter(new DoubleToRatingScaleConverter())
-              .bind(Book::getRating, Book::setRating);
+                .withConverter(new DoubleToRatingScaleConverter())
+                .bind(Book::getRating, Book::setRating);
+    }
+
+    private SerializablePredicate<LocalDate> endDatePredicate() {
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/BookForm.java:198]: Thanks for refactoring this out. It's better this way

File: src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java [hunk 6]
```diff
@@ -17,22 +17,14 @@
 
 import com.github.mvysny.kaributesting.v10.MockVaadin;
 import com.github.mvysny.kaributesting.v10.Routes;
-import com.karankumar.bookproject.backend.entity.Author;
-import com.karankumar.bookproject.backend.entity.Book;
-import com.karankumar.bookproject.backend.entity.Genre;
-import com.karankumar.bookproject.backend.entity.PredefinedShelf;
-import com.karankumar.bookproject.backend.entity.RatingScale;
+import com.karankumar.bookproject.backend.entity.*;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java:20]: Could you undo this change, please? We're not using wildcard imports

File: src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java [hunk 7]
```diff
@@ -17,22 +17,14 @@
...
 import com.vaadin.flow.component.UI;
+import com.vaadin.flow.data.binder.BinderValidationStatus;
 import com.vaadin.flow.spring.SpringServlet;
-import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.Assertions;
-import org.junit.jupiter.api.Assumptions;
-import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.*;
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java:27]: another wildcard import change to be reverted

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java (hunk 3 above)
Comment (amit1307) [reply src/main/java/com/karankumar/bookproject/ui/book/BookForm.java]: Sorry for that, Intellij default formatting did that.

File: src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java (hunk 6 above)
Comment (amit1307) [reply src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java]: Yes removed all wildcard imports.

File: src/main/java/com/karankumar/bookproject/ui/FormErrors.java (hunk 1 above)
Comment (amit1307) [reply src/main/java/com/karankumar/bookproject/ui/FormErrors.java]: Have changed the package but I was thinking of having a common errors file for all the forms.

Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/FormErrors.java]: In my view, that would make sense only if there were common constants that were needed across the form classes.

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java (hunk 3 above)
Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/ui/book/BookForm.java]: That's fine. We have an [IntelliJ code style file](https://github.com/knjk04/book-project/blob/master/ide/intellij/book_project_code_style.xml) that you can import.

File: src/main/java/com/karankumar/bookproject/ui/book/FormErrors.java [hunk 8]
```diff
@@ -0,0 +1,12 @@
+package com.karankumar.bookproject.ui.book;
+
+public class FormErrors {
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/FormErrors.java:3]: Since these errors are only for the `BookForm`, could you refactor the name to `BookFormTests` (see the comment I made about why I don't think having one class for form errors is the best idea here)

File: src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java [hunk 9]
```diff
@@ -202,16 +206,16 @@ public void formCanBeCleared() {
...
-        Assertions.assertTrue(bookForm.authorLastName.isEmpty(), "Author last name not cleared");
-        Assertions.assertTrue(bookForm.bookTitle.isEmpty(), "Book title not cleared");
-        Assertions.assertTrue(bookForm.shelf.isEmpty(), "Shelf not cleared");
-        Assertions.assertTrue(bookForm.bookGenre.isEmpty(), "Book genre not cleared");
-        Assertions.assertTrue(bookForm.pagesRead.isEmpty(), "Pages read not cleared");
-        Assertions.assertTrue(bookForm.numberOfPages.isEmpty(), "Number of pages not cleared");
-        Assertions.assertTrue(bookForm.dateStartedReading.isEmpty(), "Date started not cleared");
-        Assertions.assertTrue(bookForm.dateFinishedReading.isEmpty(), "Date finished not cleared");
-        Assertions.assertTrue(bookForm.rating.isEmpty(), "Rating not cleared");
+        assertTrue(bookForm.authorFirstName.isEmpty(), "Author first name not cleared");
```
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/ui/book/BookFormTests.java:209]: Could you change this back to use the qualified class name?

File: src/main/java/com/karankumar/bookproject/ui/FormErrors.java (hunk 1 above)
Comment (amit1307) [reply src/main/java/com/karankumar/bookproject/ui/FormErrors.java]: Sure, I had already changed the package.

**Comment by knjk04** [comment]
We haven't heard back from you in a while on this, so I just wanted to follow up. How are you getting on with the changes requested? If you need a hand, let us know over Gitter.

**Comment by amit1307** [comment]
Hey, I have addressed all the comments now. Please let me know if there is anything pending. Sorry for the delay.

File: src/main/java/com/karankumar/bookproject/ui/book/BookForm.java [hunk 10]
```diff
@@ -582,6 +583,23 @@ private void setComponentMinWidth(HasSize[] components) {
...
+    private SerializablePredicate<LocalDate> datePredicate() {
+        return startDate -> !(startDate != null && startDate.isAfter(LocalDate.now()));
+    }
+
+    private SerializablePredicate<Integer> positiveNumberPredicate() {
+        return series -> (series == null || series > 0);
+    }
+
+    private SerializablePredicate<String> authorPredicate() {
+        return firstName -> (firstName != null && !firstName.isEmpty());
```
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/ui/book/BookForm.java:600]: Since you changed both the first and last name validators, the local lambda variable would be better as `authorName`
